GetJobber/atlantis
# feat(components): Refactor/Clean Up Combobox Content

## Motivations

ComboboxContent was getting very busy and hard to read. also, we got to a place where we were getting heavy on integration tests (that's probably my fault tbh) so I wanted to see about shifting that towards unit tests where it made sense

## Changes





### Changed

- extracted the list and search subcomponents into their own files with their own styles
- added unit test files for these components, and moved any tests that were more integration or acceptance style into these where it made sense, to reduce their fragility
- reorganized the contents of `useComboboxAccessibility` and `useComboboxContent` where generic behavior of Combobox is (mostly) in the latter, and things accessibility related in the former 
- moved those hooks into their own folder and files to keep the ComboboxContent smaller and focused
- changed from using state + having it as a dependency in the `useEffect` to avoid stale values in the handler, to using a ref

https://stackoverflow.com/questions/53845595/wrong-react-hooks-behaviour-with-event-listener

## Testing

should be functionally identical. nothing should have changed at all!

I'm pretty confident in our test coverage which appears to be happy with this, but even still - give it a spin in the webstories. make sure it's doing all the things we'd expect including some of the more convoluted flows such as

- type in search to find something low on the list that you'd normally have to scroll to, select it with click/arrow keys and enter or space
- re-open the combobox, see that we're scrolled to a place where we can see our selection
- start typing a new query in search
- move the arrows down, see that it moves to the first position
- tab or click on the clear, see the list go back to normal state
- use the arrows again to move to the top of this new list
- make a selection
- see the selection is there!

**Unit test changes**

are these still valid, valuable, do you have the same level of confidence with them as unit tests? lemme know what you think!

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)
